archive/install_shield_3_z.ksy

```diff
@@ -0,0 +1,484 @@
+          so they should never be set on files in an archive.
+
+          For some reason,
+          this field apparently always has bit 7 set if no other attributes are set.
+          This seems to have no meaning and can probably be ignored.
+      - id: len_entry
+        type: u2
+        doc: Total byte size of this file entry.
+      - id: reserved_flags_1
+        type: b4
+      - id: is_uncompressed
+        type: b1
+        doc: |
+          Whether this file's data is stored uncompressed.
+          If false,
+          the data is compressed using the PKWARE DCL Implode compression algorithm.
+      - id: internal_flag
+        type: b1
+        valid: false
+        doc: |
+          According to idecomp.py,
+          this flag is used internally by InstallShield's icomp
+          and should never be true in archive files.
+      - id: has_version
+        type: b1
+        doc: |
+          May be true if this file has a version number stored in the version field.
+          This flag isn't always set reliably
+          (see also the comments in idecomp.py) -
+          it's often false even for files which have a non-zero version number stored.
+      - id: reserved_flags_2
+        type: b1
+        valid: false
+      - id: is_split
+        type: b1
+        valid:
+          expr: _root.header.is_extended or not _
+        doc: |
+          Whether this file is split across multiple archive parts.
+          Only used in extended format -
+          otherwise this field is 0
+          and the file's data is not split.
+      - id: reserved_flags_3
+        type: b7
+        valid: 0
+      - id: reserved_1
+        contents: [0x00]
+        doc: |
+          Unused according to idecomp.py.
+          Seems to be always 0.
+      - id: start_part
+        type: u1
+        valid:
+          expr: |
+            is_split
+            ? (_ > 0 and _ < end_part)
+            : _ == end_part
+        doc: |
+          Number of the split archive part in which the first part of this file's data is stored.
+          Only used in extended format -
+          otherwise this field is 0
+          and the file's data is not split.
+      - id: len_name
+        type: u1
+        valid:
+          # len_entry - (size of all fields except name)
+          eq: |
+            len_entry - (
+              end_part._sizeof
+              + directory_index._sizeof
+              + len_data_uncompressed._sizeof
+              + len_data_compressed._sizeof
+              + ofs_data._sizeof
+              + modified._sizeof
+              + attributes._sizeof
+              + len_entry._sizeof
+              + 2
+              + reserved_1._sizeof
+              + start_part._sizeof
+              + len_name._sizeof
+              + name_terminator._sizeof
+              + version._sizeof
+              + reserved_2._sizeof
+            )
+        doc: Byte size of the file name.
+      - id: name
```

Comment:
is it always a string of a particular encoding, or is it really allowed to contain arbitrary sequencies of bytes?

Reply:
This is a format from the pre-NT Windows era, so it probably doesn't have an encoding properly specified. In practice it's probably going to be code page 437, 850, or 1252, assuming a Western system. All of the test files I have are in English with pure ASCII file names, so it's difficult to say.

In cases like these I prefer to use raw byte arrays instead of hardcoding a generic encoding like ASCII or Latin-1, so that in case a file does contain unexpected non-ASCII characters, it can still be parsed using the KSY, and the application code can decide how to deal with the encoding issues (if at all). In my hacky Python script, I have a command-line option for selecting the encoding (defaults to ASCII), and if possible I avoid decoding the name/path fields at all.